### DISCUSSION TITLE
Extracting class labels

### DISCUSSION THREAD

**Discussion by FishingBehindTheNet** [onnx/onnx#5280] [Q&A]
I made an object detection model with YOLOv8 that I converted to an onnx model to make it faster on cpu. I want to extract the original label names used for annotation. I can see the labels in order in the onnx file when i open it as .txt file, but im not sure how to read this list with code. The goal is the list labels. Does anyone have a solution to get this data from the model?

**Comment by justinchuby** [comment]
Try reading the model as a protobuf file and extract information with protobuf APIs?

**Comment by cloudlakecho** [comment]
Have you tried this code:
```
import onnx
import ast

model = onnx.load(<ONNX file>)

props = { p.key : p.value for p in model.metadata_props }

if 'names' in props:
    names = ast.literal_eval(props['names'])
``` 
The user, D Dowling, posted this snippet at [Stack Overflow](https://stackoverflow.com/a/76864106/5595995). 

Although my ONNX file gives "props" as an empty List.

**Comment by FishingBehindTheNet** [comment]
UPDATE!

my current solution is perhaps a bit primitive but functional for my purpose. i read the model as a .txt file and then look for the specific line expected before it
```
            model = load(Model)
            MLabels = (
                str(model)
                .split('\nmetadata_props {\n  key: "names"\n  value: "')[-1]
                .replace("{", "")
                .replace("}", "")
                .replace("\\'", "")
                .replace("\n", "")
                .replace('"', "")
                .split(", ")
            )
            ModelLabels = ""
            for labels in MLabels:
                ModelLabels += f"{labels.split(': ')[-1]}<br>"
```

**Comment by cloudlakecho** [comment]
@FishingBehindTheNet  You mentioned that you read the ONNX file as TXT. When I see your code, you still used "load" function from "onnx" tool. Is correct? 
Or did you used "open" function from Python like `model = open(Model, mode='rb')`?

**Reply by justinchuby** [reply]
str(model)

this line should give you the string representation of the model

**Reply by cloudlakecho** [reply]
@justinchuby If ONNX file size is over 100 MB...the method:
``` 
model = onnx.load(<ONNX file>)
rep = str(model)
```
still work in Python language?

**Reply by justinchuby** [reply]
It should. Practically it may be slower

**Reply by FishingBehindTheNet** [reply]
Sorry for the late reaction. the script i'm currently using is as above:

> ```
>             model = load(Model)
>             MLabels = (
>                 str(model)
>                 .split('\nmetadata_props {\n  key: "names"\n  value: "')[-1]
>                 .replace("{", "")
>                 .replace("}", "")
>                 .replace("\\'", "")
>                 .replace("\n", "")
>                 .replace('"', "")
>                 .split(", ")
>             )
>             ModelLabels = ""
>             for labels in MLabels:
>                 ModelLabels += f"{labels.split(': ')[-1]}<br>"
> ```

the load() is from the ```from onnx import load``` that is not schown in the code snipit. I did this a while back so i'm not sure but i belief this whas the only whay i could get it to work.
